dune/grid/polyhedralgrid/grid.hh

```diff
@@ -0,0 +1,1207 @@
+            eclipseGrid->getPinchThresholdThickness() : 0.0;
+        UnstructuredGridType* cgrid = create_grid_cornerpoint(&g, z_tolerance);
+        if (!cgrid) {
+            OPM_THROW(std::runtime_error, "Failed to construct grid.");
+        }
+        return cgrid;
+    }
+
+
+  public:
+    using Base::getRealImplementation;
+
+    typedef typename Traits :: ExtraData ExtraData;
+    ExtraData extraData () const  { return this; }
+
+    template <class EntitySeed>
+    int corners( const EntitySeed& seed ) const
+    {
+      const int codim = EntitySeed :: codimension;
+      const int index = seed.index();
+      switch (codim)
+      {
+        case 0:
+          return cellVertices_[ index ].size();
+        case 1:
+          return 0;//grid_.cell_facepos[ index+1 ] - grid_.cell_facepos[ index ];
+        case dim:
+          return 1;
+      }
+      return 0;
+    }
+
+    template <class EntitySeed>
+    GlobalCoordinate
+    corner( const EntitySeed& seed, const int i ) const
+    {
+      const int codim = EntitySeed :: codimension;
+      switch (codim)
+      {
+        case 0:
+          {
+            const int coordIndex = GlobalCoordinate :: dimension * cellVertices_[ seed.index() ][ i ];
+            return copyToGlobalCoordinate( grid_.node_coordinates + coordIndex );
+          }
+        case 1:
+        {
+          const int faceVertex = grid_.face_nodes[grid_.face_nodepos[seed.index()] + i];
+          return copyToGlobalCoordinate( grid_.node_coordinates + GlobalCoordinate :: dimension * faceVertex );
+        }
+        case dim:
+          {
+            const int coordIndex = GlobalCoordinate :: dimension * seed.index();
+            return copyToGlobalCoordinate( grid_.node_coordinates + coordIndex );
+          }
+      }
+      return GlobalCoordinate( 0 );
+    }
+
+    template <class EntitySeed>
+    int subEntities( const EntitySeed& seed, const int codim ) const
+    {
+      const int index = seed.index();
+      switch (codim)
+      {
+        case 0:
+          return 1;
+        case 1:
+          return grid_.cell_facepos[ index+1 ] - grid_.cell_facepos[ index ];
+        case dim:
+          return cellVertices_[ index ].size();
+      }
+      return 0;
+    }
+
+    template <int codim>
+    typename Codim<codim>::EntitySeed
+    subEntitySeed( const typename Codim<0>::EntitySeed& elemSeed, const int i ) const
+    {
+      assert( i>= 0 && i<subEntities( elemSeed, codim ) );
+      typedef typename Codim<codim>::EntitySeed  EntitySeed;
+      if( codim == 0 )
+        return EntitySeed( elemSeed.index() );
+      else if ( codim == 1 )
+      {
+        return EntitySeed( grid_.cell_faces[ grid_.cell_facepos[ elemSeed.index() ] + i ] );
+      }
+      else if ( codim == dim )
+      {
+        return EntitySeed( cellVertices_[ elemSeed.index() ][ i ] );
+      }
+      else
+      {
+        DUNE_THROW(InvalidStateException,"codimension not availalbe");
```

Comment:
Also: why is InvalidStateException used? Would it not be better to use some invalid argument exception type? I also noticed that for some of the other methods such as `corner()` no exception is thrown, instead the origin is returned. Should that be changed to throwing for consistency?